In [31]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [32]:
(X_train,y_train),(X_test,y_test) = mnist.load_data()
X_train = X_train.reshape(60000,784)
X_test = X_test.reshape(10000,784)


In [ ]:
#Prétraitement de données
X_train =X_train.astype('float32')
X_test=X_test.astype('float32') #des nombres virgules
X_train /= 255 #mettre les valeurs dans un intervalle [0,1]
X_test /= 255
#One_Hot Encoding
nb_classes = 10
y_Train = to_categorical(y_train, nb_classes) 
y_Test = to_categorical(y_test, nb_classes) 


### Création du modèle ###
Création du modèle de régression logistique. inputshape = 784 (28*28 pixels), image représentée sous forme de vecteur.

La fonction d'activation est la fonction softmax (qui est une généralisation de la fonction logistique). Elle représente dix classes, les dix chiffre de 0 à 9.

La fonction de perte est l'entropie croisée, la métrique utilisée pour mesurer la performance du modèle est l'accuracy (nombre de prédictions correctes sur le total de prédictions, le nombre d'images)

La sortie est constituée de 10 neurone

In [34]:
model = Sequential()
model.add(Dense(10,input_shape =(784,),activation= 'softmax'))
model.compile(optimizer=SGD(learning_rate=0.05), loss='categorical_crossentropy', metrics=['accuracy'])
model.summary()

c:\Users\CYTech Student\AppData\Local\Programs\Python\Python313\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_4 (Dense)                 │ (None, 10)             │         7,850 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 7,850 (30.66 KB)

 Trainable params: 7,850 (30.66 KB)

 Non-trainable params: 0 (0.00 B)

Observer le nombres de paramètres à l'aide de la méthode summary(). Il y a 7850 = (784 + 1 (biais) x 10$ paramètres, 10 est le nombre de neurones de la couche de sortie.

In [35]:
batch_size = 128
nb_epoch = 100
history = model.fit(X_train,y_Train,epochs=nb_epoch,batch_size=batch_size, verbose=1)

Epoch 1/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8262 - loss: 0.7243
Epoch 2/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8853 - loss: 0.4318
Epoch 3/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8947 - loss: 0.3853
Epoch 4/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.9000 - loss: 0.3617
Epoch 5/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9036 - loss: 0.3470
Epoch 6/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.9065 - loss: 0.3362
Epoch 7/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.9085 - loss: 0.3282
Epoch 8/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9104 - loss: 0.3218
Epoch 9/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.9122 - loss: 0.3167
Epoch 10/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9134 - loss: 0.3122
Epoch 11/100
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.9140 - loss: 0.3085
Epoch 12/100
469/469 ━━━━━━━━━━━━━━━━━━━━

In [37]:
evaluation = model.evaluate(X_test,y_Test,verbose =1)
print("Perte pour le jeu de test est %2f, Accuracy: %2f",evaluation[0],evaluation[1])

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9242 - loss: 0.2661
Perte pour le jeu de test est %2f, Accuracy: %2f 0.2660744786262512 0.9241999983787537
